powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationBehavior.java

```diff
@@ -0,0 +1,469 @@
+
+                    final OperationEntity savedEntity = operationRepository.save(operationEntity);
+                    final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
+
+                    OperationUserActionResponse response = new OperationUserActionResponse();
+                    response.setResult(UserActionResult.APPROVED);
+                    response.setOperation(operationDetailResponse);
+                    return response;
+                } else {
+
+                    // Update failure count, check the failure count and FAIL operation if needed
+                    final Long failureCount = operationEntity.getFailureCount() + 1;
+                    final Long maxFailureCount = operationEntity.getMaxFailureCount();
+
+                    if (failureCount < maxFailureCount) {
+                        operationEntity.setFailureCount(failureCount);
+
+                        final OperationEntity savedEntity = operationRepository.save(operationEntity);
+                        final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
+
+                        OperationUserActionResponse response = new OperationUserActionResponse();
+                        response.setResult(UserActionResult.APPROVAL_FAILED);
+                        response.setOperation(operationDetailResponse);
+                        return response;
+                    } else {
+                        operationEntity.setStatus(OperationStatusDo.FAILED);
+                        operationEntity.setTimestampFinalized(currentTimestamp);
+                        operationEntity.setFailureCount(maxFailureCount); // just in case, set the failure count to max value
+
+                        final OperationEntity savedEntity = operationRepository.save(operationEntity);
+                        final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
+
+                        OperationUserActionResponse response = new OperationUserActionResponse();
+                        response.setResult(UserActionResult.OPERATION_FAILED);
+                        response.setOperation(operationDetailResponse);
+                        return response;
+                    }
+                }
+
+            } else {
+                throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_APPROVE_FAILURE);
+            }
+        } else {
+            throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_APPROVE_FAILURE);
+        }
+    }
+
+    public OperationUserActionResponse rejectOperation(OperationRejectRequest request) throws GenericServiceException {
+        final Date currentTimestamp = new Date();
+
+        final String operationId = request.getOperationId();
+        final String userId = request.getUserId();
+        final Long applicationId = request.getApplicationId();
+
+        // TODO: We might need a lock here!!!
```

Comment:
Indeed, I think that the lock would be useful here because the processing changes the operation state.

Reply:
👍 I used the lock function for all methods that modify an operation state.